### PR TITLE
Feature/nav 48 reverse arrival time direction implementation

### PR DESCRIPTION
**Pull request by clukas1**
More tests are potentially possible, but I'm pretty sure everything works as it should. And closing this issue is probably good sooner than later,

### CODE REVIEW

**Comment by clukas1** [comment]
> Thanks @clukas1, nice work!
> 
> I have some discussion points concerning the interface of Raptor and duplicated code in the scan methods.
> 
> Do you want to integrate this with the service in a separate branch?
> 
> https://github.com/naviqore/public-transit-service/blob/f6cfb360562da8c3a1d920adaec80425a995de4b/src/main/java/ch/naviqore/service/impl/PublicTransitServiceImpl.java#L62
> 
> But looks good in general.

Implementation in service coming tonight.

File: src/main/java/ch/naviqore/raptor/Raptor.java [hunk 1]
```diff
@@ -58,31 +58,86 @@ public List<Connection> routeEarliestArrival(Map<String, Integer> sourceStops, M
...
     public List<Connection> routeEarliestArrival(Map<String, Integer> sourceStops, Map<String, Integer> targetStopIds,
                                                  QueryConfig config) {
+        return route(sourceStops, targetStopIds, config, TimeType.DEPARTURE);
+    }
+
+    public List<Connection> routeLatestDeparture(String sourceStopId, String targetStopId, int arrivalTime) {
+        return routeLatestDeparture(createStopMap(sourceStopId, 0), createStopMap(targetStopId, arrivalTime));
+    }
+
+    public List<Connection> routeLatestDeparture(String sourceStopId, String targetStopId, int arrivalTime,
```
[the remark this answers is not in the document]
Comment (clukas1) [reply src/main/java/ch/naviqore/raptor/Raptor.java]: Cleaned up in 73b4ea7

File: src/main/java/ch/naviqore/raptor/Raptor.java [hunk 2]
```diff
@@ -58,31 +58,86 @@ public List<Connection> routeEarliestArrival(Map<String, Integer> sourceStops, M
...
+                                                 QueryConfig config) {
+        return routeLatestDeparture(createStopMap(sourceStopId, 0), createStopMap(targetStopId, arrivalTime), config);
+    }
+
+    public List<Connection> routeLatestDeparture(Map<String, Integer> sourceStops, Map<String, Integer> targetStops) {
```
[the remark this answers is not in the document]
Comment (clukas1) [reply src/main/java/ch/naviqore/raptor/Raptor.java]: Simplified even more 73b4ea7

File: src/main/java/ch/naviqore/raptor/Raptor.java [hunk 3]
```diff
@@ -58,31 +58,86 @@ public List<Connection> routeEarliestArrival(Map<String, Integer> sourceStops, M
...
 
-        log.info("Routing earliest arrival from {} to {} at {}", sourceStopIdxs, targetStopIdxs, departureTimes);
-        List<Leg[]> earliestArrivalsPerRound = spawnFromSourceStop(sourceStopIdxs, targetStopIdxs, departureTimes,
-                walkingDurationsToTarget, config);
+        List<Leg[]> earliestArrivalsPerRound = spawnFromStop(sourceStopIdxs, targetStopIdxs, departureTimes,
+                walkingDurationsToTarget, config, timeType);
 
         // get pareto-optimal solutions
-        return reconstructParetoOptimalSolutions(earliestArrivalsPerRound, targetStopIdxs);
+        if (timeType == TimeType.DEPARTURE) {
```
[the remark this answers is not in the document]
Comment (clukas1) [reply src/main/java/ch/naviqore/raptor/Raptor.java]: 84b484a

File: src/main/java/ch/naviqore/raptor/Raptor.java [hunk 4]
```diff
@@ -143,167 +201,417 @@ private List<Leg[]> spawnFromSourceStop(int[] sourceStopIdxs, int[] targetStopId
...
+     *
+     * @param currentRouteIdx           - The index of the current route.
+     * @param earliestArrivals          - The earliest arrival time for each stop.
+     * @param earliestArrivalsLastRound - The earliest arrival leg for each stop in the last round.
+     * @param earliestArrivalsThisRound - The earliest arrival leg for each stop in the current round.
+     * @param markedStops               - The set of marked stops from the previous round.
+     * @param markedStopsNext           - The set of marked stops for the next round.
+     * @param minTransferDuration       - The minimum transfer duration time.
+     */
+    private void scanRouteForward(int currentRouteIdx, int[] earliestArrivals, Leg[] earliestArrivalsLastRound,
```
[the remark this answers is not in the document]
Comment (clukas1) [reply src/main/java/ch/naviqore/raptor/Raptor.java]: I was struggling with this one, should ask ChatGPT more often :)

84b484a

File: src/main/java/ch/naviqore/raptor/Raptor.java [hunk 5]
```diff
@@ -109,29 +164,32 @@ public Map<String, Connection> getIsoLines(Map<String, Integer> sourceStops, Que
...
-            throw new IllegalArgumentException("Source stops and departure times must have the same size.");
+        if (sourceStopIdxs.length != sourceTimes.length) {
+            throw new IllegalArgumentException("Source stops and departure/arrival times must have the same size.");
         }
 
         if (targetStopIdxs.length != walkingDurationsToTarget.length) {
             throw new IllegalArgumentException("Target stops and walking durations to target must have the same size.");
         }
 
         // This is used to determine the criteria for maximum travel time
```
[the remark this answers is not in the document]
Comment (clukas1) [reply src/main/java/ch/naviqore/raptor/Raptor.java]: 84b484a

**Comment by clukas1** [comment]
@munterfi All comments (and more have been addressed), including the Service Implementation (also on Viewer side, but haven't created pull request yet on viewer repo)

All unit tests etc. pass, but I get some weird behavior, which I still want to debug.
image

And sometimes (maybe already resolved) the legs not connected error. So more debugging. Feel free to dive into it, but will have some more time on Tuesday or Wednesday to follow up.

**Review by munterfi — CHANGES_REQUESTED** [review]
Many thanks for your efforts, @clukas1!

Looks good in general.

When I run the service with the test feed (`src/test/resources/ch/naviqore/gtfs/schedule/sample-feed-1.zip`) and execute the example requests in `ApplicationTest.rest` there are some strange errors and null pointers.

## Connection

- Request:

```
### Connection between two stops
GET http://localhost:8080/routing/connections?sourceStopId=STAGECOACH&targetStopId=NANAA&
    departureDateTime=2010-01-01T01:10:01
Accept: application/json
```

- Response:
```
GET http://localhost:8080/routing/connections?sourceStopId=STAGECOACH&targetStopId=NANAA&departureDateTime=2010-01-01T01:10:01

HTTP/1.1 500 
Content-Type: application/json
Transfer-Encoding: chunked
Date: Wed, 19 Jun 2024 08:16:53 GMT
Connection: close

{
  "timestamp": "2024-06-19T08:16:53.668+00:00",
  "status": 500,
  "error": "Internal Server Error",
  "path": "/routing/connections"
}
Response file saved.
> 2024-06-19T101653.500.json

Response code: 500; Time: 16ms (16 ms); Content length: 120 bytes (120 B)
```

- Error:

```
[WARN ] 2024-06-19 10:16:53.668 [http-nio-8080-exec-5] Raptor - Stop ID STAGECOACH not found in lookup removing from query.
2024-06-19T10:16:53.668+02:00 ERROR 13844 --- [nio-8080-exec-5] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: java.lang.IllegalArgumentException: No valid stops provided.] with root cause

java.lang.IllegalArgumentException: No valid stops provided.
	at ch.naviqore.raptor.Raptor$InputValidator.validateStops(Raptor.java:774) ~[classes/:na]
	at ch.naviqore.raptor.Raptor.route(Raptor.java:71) ~[classes/:na]
	at ch.naviqore.service.impl.PublicTransitServiceImpl.getConnections(PublicTransitServiceImpl.java:189) ~[classes/:na]
	at ch.naviqore.service.impl.PublicTransitServiceImpl.getConnections(PublicTransitServiceImpl.java:130) ~[classes/:na]
	at ch.naviqore.app.service.PublicTransitSpringService.getConnections(PublicTransitSpringService.java:80) ~[classes/:na]
	at ch.naviqore.app.controller.RoutingController.getConnections(RoutingController.java:113) ~[classes/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[na:na]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:255) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:188) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:926) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:831) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564) ~[tomcat-embed-core-10.1.24.jar:6.0]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[tomcat-embed-core-10.1.24.jar:6.0]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[tomcat-embed-websocket-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.8.jar:6.1.8]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.8.jar:6.1.8]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.8.jar:6.1.8]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:389) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:896) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
```

## Isoline

And also the isoline request fails:

- Request:
```
### Isoline from stop
GET http://localhost:8080/routing/isolines?sourceStopId=STAGECOACH&maxTransferNumber=3&maxTravelTime=7200&
    departureDateTime=2010-01-01T01:10:01
Accept: application/json
```

- Response:
```
GET http://localhost:8080/routing/isolines?sourceStopId=STAGECOACH&maxTransferNumber=3&maxTravelTime=7200&departureDateTime=2010-01-01T01:10:01

HTTP/1.1 500 
Content-Type: application/json
Transfer-Encoding: chunked
Date: Wed, 19 Jun 2024 08:20:51 GMT
Connection: close

{
  "timestamp": "2024-06-19T08:20:51.472+00:00",
  "status": 500,
  "error": "Internal Server Error",
  "path": "/routing/isolines"
}
Response file saved.
> 2024-06-19T102051.500.json

Response code: 500; Time: 0ms (0 ms); Content length: 117 bytes (117 B)
```

- Error:
```
2024-06-19T10:20:51.472+02:00 ERROR 13844 --- [io-8080-exec-10] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed: java.lang.NullPointerException: Cannot invoke "java.time.LocalDateTime.toLocalTime()" because "time" is null] with root cause

java.lang.NullPointerException: Cannot invoke "java.time.LocalDateTime.toLocalTime()" because "time" is null
	at ch.naviqore.service.impl.PublicTransitServiceImpl.getIsoLines(PublicTransitServiceImpl.java:265) ~[classes/:na]
	at ch.naviqore.app.service.PublicTransitSpringService.getIsoLines(PublicTransitSpringService.java:104) ~[classes/:na]
	at ch.naviqore.app.controller.RoutingController.getIsolines(RoutingController.java:156) ~[classes/:na]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[na:na]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:255) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:188) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:926) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:831) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:903) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:564) ~[tomcat-embed-core-10.1.24.jar:6.0]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885) ~[spring-webmvc-6.1.8.jar:6.1.8]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[tomcat-embed-core-10.1.24.jar:6.0]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[tomcat-embed-websocket-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.8.jar:6.1.8]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.8.jar:6.1.8]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-6.1.8.jar:6.1.8]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.1.8.jar:6.1.8]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:389) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:896) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1741) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63) ~[tomcat-embed-core-10.1.24.jar:10.1.24]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
```

I think we should refactor the Raptor as soon as possible and not add any new features during this time. Because at the moment I've lost track of what's happening when and where, which makes it very difficult for me to fix bugs and debug. 😅

**Comment by munterfi** [comment]
Ok, the isoline error is due to a change of the name from `departureDateTime` to `dateTime`:

This works:

```
### Connection between two stops
GET http://localhost:8080/routing/connections?sourceStopId=STAGECOACH&targetStopId=NANAA&
    departureDateTime=2010-01-01T01:10:01
Accept: application/json
```

Nevertheless, we should set the default time to now if no time is specified.

**Comment by munterfi** [comment]
I fixed the errors in https://github.com/naviqore/public-transit-service/pull/64/commits/b8b801d3ec8075bf37d5385acdff304f93cee302.

The error in connection is due to the limited validity of the example feed. If we don't provide a `dateTime`, the current time is used. The example schedule is only valid until 2010, which results in no departing trips. However, the error message is a bit misleading: "No valid stops provided."

The stop is valid, but there are no departures. We should introduce better error messages and also pass them to the user via the REST Controller instead of a generic 500 - Internal Server Error.

I created a new issue, NAV-107, for this.

**Review by munterfi — APPROVED** [review]

**Pull request merged**